packages/server/src/sdk/app/applications/sync.ts

```diff
@@ -1,6 +1,147 @@
+    return
+  }
+  await context.doInAppContext(appId, async () => {
+    const db = context.getAppDB()
+    for (let user of users) {
+      let ctxUser = user as ContextUser
+      let deletedUser = false
+      const metadataId = generateUserMetadataID(user._id!)
+      if ((user as DeletedUser).deleted) {
+        deletedUser = true
+      }
+
+      // make sure role is correct
+      if (!deletedUser) {
+        ctxUser = await processUser(ctxUser, { appId, groups })
+      }
+      let roleId = ctxUser.roleId
+      if (roleId === roles.BUILTIN_ROLE_IDS.PUBLIC) {
+        roleId = undefined
+      }
+
+      let metadata
+      try {
+        metadata = await db.get(metadataId)
+      } catch (err: any) {
+        if (err.status !== 404) {
+          throw err
+        }
+        // no metadata and user is to be deleted, can skip
+        // no role - user isn't in app anyway
+        if (!roleId) {
+          continue
+        } else if (!deletedUser) {
+          // doesn't exist yet, creating it
+          metadata = {
+            tableId: InternalTables.USER_METADATA,
+          }
+        }
+      }
+
+      // the user doesn't exist, or doesn't have a role anymore
+      // get rid of their metadata
+      if (deletedUser || !roleId) {
+        await db.remove(metadata)
+        continue
+      }
+
+      // assign the roleId for the metadata doc
+      if (roleId) {
+        metadata.roleId = roleId
+      }
+
+      let combined = sdk.users.combineMetadataAndUser(ctxUser, metadata)
+      // if no combined returned, there are no updates to make
+      if (combined) {
+        await db.put(combined)
+      }
+    }
+  })
+}
+
+async function syncUsersToAllApps(userIds: string[]) {
+  // list of users, if one has been deleted it will be undefined in array
+  const users = (await getGlobalUsers(userIds, {
+    noProcessing: true,
+  })) as User[]
+  const groups = await proSdk.groups.fetch()
+  const finalUsers: (User | DeletedUser)[] = []
+  for (let userId of userIds) {
+    const user = users.find(user => user._id === userId)
+    if (!user) {
+      finalUsers.push({ _id: userId, deleted: true })
+    } else {
+      finalUsers.push(user)
+    }
+  }
+  const devAppIds = await dbCore.getDevAppIDs()
+  let promises = []
+  for (let devAppId of devAppIds) {
+    const prodAppId = dbCore.getProdAppID(devAppId)
+    for (let appId of [prodAppId, devAppId]) {
+      promises.push(syncUsersToApp(appId, finalUsers, groups))
+    }
+  }
+  const resp = await Promise.allSettled(promises)
+  const failed = resp.filter(promise => promise.status === "rejected")
+  if (failed.length > 0) {
+    const reasons = failed.map(fail => (fail as PromiseRejectedResult).reason)
+    logging.logAlert("Failed to sync users to apps", reasons)
+  }
+}
+
+export function initUserGroupSync(updateCb?: (docId: string) => void) {
+  const types = [constants.DocumentType.USER, constants.DocumentType.GROUP]
+  docUpdates.process(types, async update => {
```

Comment:
I think this is maybe the wrong place to put this, with the event intg being nested inside the SDK. It feels like this could be a candidate to pull out to a top level `events` package in server. 

Ideally we could init event processing in a single place in the server for all future events. 
e.g. 

```ts
class BullEventsProcessor implements EventProcessor {

  processors: { [key in Event]: EventProcessor[] }

  constructor(processors) {
      this.processors = processors
  }

  init() {
      // hand off to the event processor interface
     queue.process(job: EventPayload, () => { await processEvent (job.event, job.identity... })
  }
 
  // main process impl to hand off to individual processors
  async processEvent(
    event: Event,
    identity: Identity,
    properties: any,
    timestamp?: string
  ): Promise<void> {
    // process the event with any registered sub-processors
     const processors =  processors[event]
     for (processor in processors) {
       processor.processEvent(event, identity, properties, timestamp)
     }
  }
```

We can then create a dedicated processor that can handle user group sync and register it with the above 
```ts
class UserGroupSyncEventProcessor implements EventProcessor {}
```

This will set us up for really easy and clear expansion, by just implementing `EventProcessor` and registering it with the main processor.